File: src/test/java/com/github/alexandrecarlton/idea/settings/applier/impl/tools/file_watchers/FileWatcherSettingsApplierTest.java
package com.github.alexandrecarlton.idea.settings.applier.impl.tools.file_watchers;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

import com.github.alexandrecarlton.idea.settings.applier.api.SettingsApplier;
import com.github.alexandrecarlton.idea.settings.applier.impl.common.FileTypeMapper;
import com.github.alexandrecarlton.idea.settings.fixtures.IdeaSettingsTestFixture;
import com.github.alexandrecarlton.idea.settings.layout.common.FileType;
import com.github.alexandrecarlton.idea.settings.layout.tools.file_watchers.FileWatcherSettings;
import com.github.alexandrecarlton.idea.settings.layout.tools.file_watchers.ImmutableFileWatcherAdvancedOptionsSettings;
import com.github.alexandrecarlton.idea.settings.layout.tools.file_watchers.ImmutableFileWatcherFilesToWatchSettings;
import com.github.alexandrecarlton.idea.settings.layout.tools.file_watchers.ImmutableFileWatcherSettings;
import com.github.alexandrecarlton.idea.settings.layout.tools.file_watchers.ImmutableFileWatcherToolToRunOnChangesSettings;
import com.intellij.lang.javascript.JavaScriptFileType;
import com.intellij.plugins.watcher.model.ProjectTasksOptions;
import com.intellij.plugins.watcher.model.TaskOptions;
import org.junit.Before;
import org.junit.Test;

public class FileWatcherSettingsApplierTest extends IdeaSettingsTestFixture {

  private SettingsApplier<FileWatcherSettings> settingsApplier;
  private ProjectTasksOptions projectTasksOptions;
  private FileTypeMapper fileTypeMapper;

  @Before
  public void setUp() {
    fileTypeMapper = mock(FileTypeMapper.class);
    projectTasksOptions = ProjectTasksOptions.getInstance(project);
    settingsApplier = new FileWatcherSettingsApplier(projectTasksOptions, fileTypeMapper);
  }

  @Test
  public void basicSettingsApplied() {
    when(fileTypeMapper.mapFileType(FileType.JAVASCRIPT)).thenReturn(JavaScriptFileType.INSTANCE);

    settingsApplier.apply(ImmutableFileWatcherSettings.builder()
      .name("Basic File Watcher")
      .toolToRunOnChanges(ImmutableFileWatcherToolToRunOnChangesSettings.builder()
          .program("/usr/bin/foo")
          .arguments("bar")
          .outputPathsToRefresh("$FilePath$")
          .build())
      .filesToWatch(ImmutableFileWatcherFilesToWatchSettings.builder()
          .fileType(FileType.JAVASCRIPT)
          .scope("Module 'foo'")
          .build())
      .build());

    TaskOptions taskOptions = projectTasksOptions.getTasks()
      .stream()
      .filter(pair -> "Basic File Watcher".equals(pair.first.getName()))
      .findFirst()
      .get()
      .first;
    assertThat(taskOptions.getName()).isEqualTo("Basic File Watcher");
    assertThat(taskOptions.getProgram()).isEqualTo("/usr/bin/foo");
    assertThat(taskOptions.getArguments()).isEqualTo("bar");
    assertThat(taskOptions.getOutput()).isEqualTo("$FilePath$");
    assertThat(taskOptions.getFileExtension()).isEqualTo("js");
    assertThat(taskOptions.getScopeName()).isEqualTo("Module 'foo'");
  }

  @Test
  public void advancedSettingsApplied() {
    settingsApplier.apply(ImmutableFileWatcherSettings.builder()
      .name("Advanced Watcher Settings")
      .toolToRunOnChanges(ImmutableFileWatcherToolToRunOnChangesSettings.builder()
        .program("/usr/bin/foo")
        .arguments("bar")
        .outputPathsToRefresh("$FilePath$")
        .build())
      .filesToWatch(ImmutableFileWatcherFilesToWatchSettings.builder()
        .fileType(FileType.JAVASCRIPT)
        .scope("Module 'foo'")
        .build())
      .advancedOptions(ImmutableFileWatcherAdvancedOptionsSettings.builder()
        .autoSaveEditedFilesToTriggerTheWatcher(true)
        .triggerTheWatcherOnExternalChanges(false)
        .triggerTheWatcherRegardlessOfSyntaxErrors(true)
        .build())
      .build());

    TaskOptions taskOptions = projectTasksOptions.getTasks()
      .stream()
      .filter(pair -> "Advanced Watcher Settings".equals(pair.first.getName()))
      .findFirst()
      .get()
      .first;
    assertThat(taskOptions.isImmediateSync()).isTrue();
    assertThat(taskOptions.isRunOnExternalChanges()).isFalse();
    assertThat(taskOptions.isCheckSyntaxErrors()).isTrue();

  }

}

File: src/test/kotlin/com/github/alexandrecarlton/idea/settings/applier/impl/tools/file_watchers/FileWatcherSettingsApplierTest.kt
package com.github.alexandrecarlton.idea.settings.applier.impl.tools.file_watchers

import com.github.alexandrecarlton.idea.settings.applier.api.SettingsApplier
import com.github.alexandrecarlton.idea.settings.applier.impl.common.FileTypeMapper
import com.github.alexandrecarlton.idea.settings.fixtures.IdeaSettingsTestFixture
import com.github.alexandrecarlton.idea.settings.layout.common.FileType
import com.github.alexandrecarlton.idea.settings.layout.tools.file_watchers.FileWatcherAdvancedOptionsSettings
import com.github.alexandrecarlton.idea.settings.layout.tools.file_watchers.FileWatcherFilesToWatchSettings
import com.github.alexandrecarlton.idea.settings.layout.tools.file_watchers.FileWatcherSettings
import com.github.alexandrecarlton.idea.settings.layout.tools.file_watchers.FileWatcherToolToRunOnChangesSettings
import com.intellij.lang.javascript.JavaScriptFileType
import com.intellij.plugins.watcher.model.ProjectTasksOptions
import org.assertj.core.api.Assertions.assertThat
import org.junit.Before
import org.junit.Test
import java.util.function.Supplier

class FileWatcherSettingsApplierTest : IdeaSettingsTestFixture() {

    private lateinit var settingsApplier: SettingsApplier<FileWatcherSettings>
    private lateinit var projectTasksOptions: ProjectTasksOptions
    private val fileTypeMapper = FileTypeMapper(Supplier { JavaScriptFileType.INSTANCE })

    @Before
    public override fun setUp() {
        projectTasksOptions = ProjectTasksOptions.getInstance(project)
        settingsApplier = FileWatcherSettingsApplier(projectTasksOptions, fileTypeMapper)
    }

    @Test
    fun basicSettingsApplied() {
        settingsApplier.apply(FileWatcherSettings(
            name = "Basic File Watcher",
            filesToWatch = FileWatcherFilesToWatchSettings(
                fileType = FileType.JAVASCRIPT,
                scope = "Module 'foo'"),
            toolToRunOnChanges = FileWatcherToolToRunOnChangesSettings(
                program = "/usr/bin/foo",
                arguments = "bar",
                outputPathsToRefresh = "\$FilePath$",
                workingDirectory = "\$ProjectFileDir$"
                )))

        val taskOptions = projectTasksOptions.tasks
                .stream()
                .filter { pair -> "Basic File Watcher" == pair.first.name }
                .findFirst()
                .get()
                .first
        assertThat(taskOptions.name).isEqualTo("Basic File Watcher")
        assertThat(taskOptions.program).isEqualTo("/usr/bin/foo")
        assertThat(taskOptions.arguments).isEqualTo("bar")
        assertThat(taskOptions.output).isEqualTo("\$FilePath$")
        assertThat(taskOptions.workingDir).isEqualTo("\$ProjectFileDir$")
        assertThat(taskOptions.fileExtension).isEqualTo("js")
        assertThat(taskOptions.scopeName).isEqualTo("Module 'foo'")
    }

    @Test
    fun advancedSettingsApplied() {
        settingsApplier.apply(FileWatcherSettings(
            name = "Advanced Watcher Settings",
            filesToWatch = FileWatcherFilesToWatchSettings(
                fileType = FileType.JAVASCRIPT,
                scope = "Module 'foo'"),
            toolToRunOnChanges = FileWatcherToolToRunOnChangesSettings(
                program = "/usr/bin/foo",
                arguments = "bar",
                outputPathsToRefresh = "\$FilePath"),
            advancedOptions = FileWatcherAdvancedOptionsSettings(
                autoSaveEditedFilesToTriggerTheWatcher = true,
                triggerTheWatcherOnExternalChanges = false,
                triggerTheWatcherRegardlessOfSyntaxErrors = true)))

        val taskOptions = projectTasksOptions.tasks
                .stream()
                .filter { pair -> "Advanced Watcher Settings" == pair.first.name }
                .findFirst()
                .get()
                .first
        assertThat(taskOptions.isImmediateSync).isTrue()
        assertThat(taskOptions.isRunOnExternalChanges).isFalse()
        assertThat(taskOptions.isCheckSyntaxErrors).isFalse()
    }
}
